modules/platform/nuxeo-platform-web-common/src/main/java/org/nuxeo/ecm/platform/web/common/idempotency/NuxeoIdempotentFilter.java

```diff
@@ -0,0 +1,168 @@
+ */
+package org.nuxeo.ecm.platform.web.common.idempotency;
+
+import java.io.IOException;
+import java.time.Duration;
+import java.util.Set;
+
+import javax.servlet.Filter;
+import javax.servlet.FilterChain;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import org.apache.http.client.methods.HttpDelete;
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.client.methods.HttpHead;
+import org.apache.http.client.methods.HttpOptions;
+import org.apache.http.client.methods.HttpPut;
+import org.apache.http.client.methods.HttpTrace;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.nuxeo.runtime.api.Framework;
+import org.nuxeo.runtime.kv.KeyValueService;
+import org.nuxeo.runtime.kv.KeyValueStore;
+import org.nuxeo.runtime.services.config.ConfigurationService;
+
+/**
+ * Filter handling an idempotency key in POST requests.
+ * <p>
+ * If {@link #HEADER_KEY} is found in the request header, will intercept request handling to:
+ * <ul>
+ * <li>mark the request as being processed
+ * <li>capture the response when request was processed without any error and store it
+ * <li>return the stored response if a subsequent request with the same key is processed again
+ * <li>return a conflict response if a request with the same key is processed while the first request is still in
+ * progress.
+ * </ul>
+ *
+ * @since 11.5
+ */
+public class NuxeoIdempotentFilter implements Filter {
+
+    private static final Logger log = LogManager.getLogger(NuxeoIdempotentFilter.class);
+
+    public static final String HEADER_KEY = "Idempotency-Key";
+
+    public static final String STORE_PROPERTY = "org.nuxeo.request.idempotency.keyvaluestore.name";
+
+    public static final String DEFAULT_STORE = "idempotentrequest";
+
+    protected static final long DEFAULT_TTL_SECONDS = Duration.ofDays(1).toSeconds();
+
+    public static final String TTL_SECONDS_PROPERTY = "org.nuxeo.request.idempotency.ttl.seconds";
+
+    public static final String INPROGRESS_MARKER = "IDEMPOTENCY_INPROGRESS_MARKER";
+
+    public static final String INFO_SUFFIX = "_info";
+
+    protected static final int DEFERRED_OUTPUT_STREAM_THRESHOLD = 1024 * 1024; // 1 MB
+
+    // safe methods according to RFC 7231 4.2.1
+    protected static final Set<String> SAFE_METHODS = Set.of(HttpGet.METHOD_NAME, HttpHead.METHOD_NAME,
+            HttpOptions.METHOD_NAME, HttpTrace.METHOD_NAME);
+
+    // idempotent methods according to RFC 7231 4.2.2
+    protected static final Set<String> IDEMPOTENT_METHODS = Set.of(HttpPut.METHOD_NAME, HttpDelete.METHOD_NAME);
+
+    protected long getTTLSeconds() {
+        ConfigurationService cs = Framework.getService(ConfigurationService.class);
+        if (cs != null) {
+            return cs.getLong(TTL_SECONDS_PROPERTY, DEFAULT_TTL_SECONDS);
```

Comment:
Probably better to use `getDuration` here, it's more flexible for users. And then convert to seconds. And rename the property to not have "seconds" in the name or value. Then `DEFAULT_TTL_SECONDS` can be just a duration too, and you only convert to seconds at the last moment: `long ttl = getTTLSeconds()` -> `long ttl = getTTL().toSeconds()`.